modules/dnn/src/onnx/onnx_importer.cpp

```diff
@@ -537,6 +537,29 @@ void ONNXImporter::populateNet(Net dstNet)
         {
             replaceLayerParam(layerParams, "size", "local_size");
         }
+        else if (layer_type == "InstanceNormalization")
+        {
+            if (node_proto.input_size() != 3)
+                CV_Error(Error::StsNotImplemented,
+                         "Expected input, scale, bias");
+
+            layerParams.type = "InstanceNorm";
+            replaceLayerParam(layerParams, "epsilon", "eps");
+
+            if (!node_proto.input(1).empty()) {
+                layerParams.set("has_weight", true);
+                layerParams.blobs.push_back(getBlob(node_proto, constBlobs, 1));  // weightData
+            } else {
+                layerParams.set("has_weight", false);
+            }
+
+            if (!node_proto.input(2).empty()) {
+                layerParams.set("has_bias", true);
+                layerParams.blobs.push_back(getBlob(node_proto, constBlobs, 2)); // biasData
+            } else {
+                layerParams.set("has_bias", false);
+            }
```

Comment:
From our experience with IntanceNorm layer from Torch (see [torch_importer.cpp](https://github.com/opencv/opencv/blob/master/modules/dnn/src/torch/torch_importer.cpp)) and from TensorFlow (see [tf_importer.cpp](https://github.com/opencv/opencv/blob/master/modules/dnn/src/tensorflow/tf_importer.cpp) at FusedBatchNorm with training mode) we can use existing MVN and BatchNorm layer. So there is no need to add an extra layer but add two existing layers in a row.

Reply:
Hey @dkurt thanks for your answer. So I should connect BatchNorm->MVN layers in the onnx_importer when layer_type == "InstanceNormalization". 
Could you suggest how should I go about this. Should I create a Pin struct as in [tf_importer](https://github.com/opencv/opencv/blob/f33f88de3138366bc370f5d49d08d351179b910d/modules/dnn/src/tensorflow/tf_importer.cpp#L61)?
Couldnt I just keep the new layer and implement the BN operation and then the MVN operation inside it?

Reply:
That's a god question actually because it'll be the first time we do it for ONNX importer.
Try to create a layer MVN with some prefixed name and connect it to origin input. Then you need to change base batch norm layer's input to MVN's name.


origin network:
```
inp_1 -> InstanceNorm
```

we will create:
```
inp_1 -> MVN -> BatchNorm
```

Reply:
Hey @dkurt I am trying to do what you suggested but I am  a bit lost. The connections and layer additions for Onnx seem to happen [here](https://github.com/opencv/opencv/blob/974491f93d0faa35bc92eef2d85d6b14ef4a40ae/modules/dnn/src/onnx/onnx_importer.cpp#L845), so how am I supposed to connect the layers inside the layer_type == "InstanceNormalization" condition?

I tried something along these lines

``` 
else if (layer_type == "InstanceNormalization")
 {
      if (node_proto.input_size() != 3)
          CV_Error(Error::StsNotImplemented,
                         "Expected input, scale, bias");

            layerParams.type = "InstanceNorm";
            replaceLayerParam(layerParams, "epsilon", "eps");

            if (!node_proto.input(1).empty()) {
                layerParams.set("has_weight", true);
                layerParams.blobs.push_back(getBlob(node_proto, constBlobs, 1));  // weightData
            } else {
                layerParams.set("has_weight", false);
            }

            if (!node_proto.input(2).empty()) {
                layerParams.set("has_bias", true);
                layerParams.blobs.push_back(getBlob(node_proto, constBlobs, 2)); // biasData
            } else {
                layerParams.set("has_bias", false);
            }           

           std::string mvnName = layerParams.name + "/MVN";           
           CV_Assert(layer_id.find(layerParams.name) == layer_id.end());
           int mvnId = dstNet.addLayer(mvnName, "MVN", layerParams);           
           layer_id.insert(std::make_pair(layerParams.name, LayerInfo(mvnId, 0)));
           layerId = layer_id.find(node_proto.input(0)); 
           dstNet.connect(layerId->second.layerId, layerId->second.outputId, mvnId, 0);           
           layerParams.type = "BatchNorm"; 
           layerParams.name = layerParams.name + "/BatchNorm";
```
This, of course, did not work `(-215:Assertion failed) shapeIt != outShapes.end()`.  Am I suppose to add and connect the BatchNorm layer right after those lines? If so what is [this loop](https://github.com/opencv/opencv/blob/974491f93d0faa35bc92eef2d85d6b14ef4a40ae/modules/dnn/src/onnx/onnx_importer.cpp#L845) doing. Once more, thanks for your help!! :)

Reply:
MVN is a separate layer and you need to use separate LayerParams for it

Reply:
Ok I will do that. Could you please answer the questions I posted before so I can get some idea on how to connect these 2 layers?

Reply:
Hey @dkurt , so I am trying to do what you suggested, and I cant connect the MVN to the BatchNorm layer. This is how I connect the Input to and MVN layer, and this seems to work fine, however I cannot connect the MVN layer to the BN layer. This is the closest I have gotten:
```            
        else if (layer_type == "InstanceNormalization")
        {
            if (node_proto.input_size() != 3)
                CV_Error(Error::StsNotImplemented,
                         "Expected input, scale, bias")
            
            //prevents connecting these layers again at the end of the function since we will connect them here         
            layerParams.composed = true;     
                              
            layerParams.blobs.resize(2);  

            if (!node_proto.input(1).empty()) {
                layerParams.blobs.resize(layerParams.blobs.size() + 1);
                layerParams.set("has_weight", true);
                layerParams.blobs.back() = getBlob(node_proto, constBlobs, 1);  // weightData
            } else {
                layerParams.set("has_weight", false);
            }

            if (!node_proto.input(2).empty()) {
                layerParams.blobs.resize(layerParams.blobs.size() + 1);
                layerParams.set("has_bias", true);
                layerParams.blobs.back() = getBlob(node_proto, constBlobs, 2);  // biasData
            } else {
                layerParams.set("has_bias", false);
            } 

          Mat mean, std;                                                                                                                                                                
          mean = Mat::zeros(1, layerParams.blobs[2].total(), CV_32F);                                                                                                                   
          std = Mat::ones(1, layerParams.blobs[2].total(), CV_32F);                                                                                                                     
          layerParams.blobs[0] = mean;                                                                                                                                                  
          layerParams.blobs[1] = std;                                                                                                                                                                                                                                                           
          
          LayerParams mvnParams;                                                                                                                                                        
          mvnParams.name = layerParams.name + "/MVN";                                                                                                                                   
          mvnParams.type = "MVN";                                                                                                                                                       
                                                                                                                                                                                    
           CV_Assert(layer_id.find(mvnParams.name) == layer_id.end());                                                                                                                   
           int mvnId = dstNet.addLayer(mvnParams.name, mvnParams.type, mvnParams);                                                                                                       
           layer_id.insert(std::make_pair(mvnParams.name, LayerInfo(mvnId, 0)));                                                                                                         
                                                                                                                                                                                       
           std::vector<MatShape> layerInpShapes, layerOutShapes, layerInternalShapes;                                                                                                    
           for (int j = 0; j < node_proto.input_size(); j++) {                                                                                                                           
                       layerId = layer_id.find(node_proto.input(j));                                                                                                                             
                       if (layerId != layer_id.end()) {                                                                                                                                          
                            dstNet.connect(layerId->second.layerId, layerId->second.outputId, mvnId, j);                                                                                          
                            shapeIt = outShapes.find(node_proto.input(j));
                            CV_Assert(shapeIt != outShapes.end());
                            layerInpShapes.push_back(shapeIt->second);
                      }
            }

            // Compute shape of output blob for this layer.                                                                                                                                   
            Ptr<Layer> layer = dstNet.getLayer(mvnId);                                                                                                                                        
            layer->getMemoryShapes(layerInpShapes, 0, layerOutShapes, layerInternalShapes);                                                                                                   
            CV_Assert(!layerOutShapes.empty());                                                                                                                                               
           outShapes[mvnParams.name] = layerOutShapes[0]; 

           layerParams.type = "BatchNorm";                                                                                                                                                   
           layerParams.name += "/BatchNorm";                                                                                                                                                                                                                                                                                                                             
           int id = dstNet.addLayer(layerParams.name, layerParams.type, layerParams);                                                                                                        
           layer_id.insert(std::make_pair(layerParams.name, LayerInfo(id, 0)));                                                                                                              
           std::vector<MatShape> layerInpShapes2, layerOutShapes2, layerInternalShapes2;                                                                                                     
                                                                                                                                                                                            
             for (int j = 0; j < node_proto.input_size(); j++) {                                                                                                                           
                       layerId = layer_id.find(node_proto.input(j));                                                                                                                             
                       if (layerId != layer_id.end()) {                                                                                                                                          
                            dstNet.connect(layerId->second.layerId + 1, layerId->second.outputId, id, j);                                                                                          
                            shapeIt = outShapes.find(node_proto.input(j));
                            CV_Assert(shapeIt != outShapes.end());
                            layerInpShapes2.push_back(shapeIt->second);
                      }
            }                                                                                                                                               
            // Compute shape of output blob for this layer.                                                                                                                                   
            Ptr<Layer> layer = dstNet.getLayer(id);                                                                                                                                        
            layer->getMemoryShapes(layerInpShapes2, 0, layerOutShapes2, layerInternalShapes2);                                                                                                   
            CV_Assert(!layerOutShapes2.empty());                                                                                                                                               
           outShapes[layerParams.name] = layerOutShapes2[0]; 
                                                                                                                        
            } 
``` 
Do you have any ideas on how these 2 layers can be connected? Thanks in advance

Reply:
Hey @dkurt I finally managed to make the layer connection as you suggested (inp1->MVN ->BN). However... the output is not even remotely closed as to what to PyTorch implementation of instanceNorm yields. Are you sure that MVN + BN = InstanceNorm? 
I checked and both layers run fine with their expected input and output, however the results differ A LOT. 

Should I make a commit so you can see what I have changed?

Reply:
We expect that no InstanceNormalization layer will be added but in importer you can reuse existing MVN and batch norm. So please share the code.

Reply:
Done :)